addons/sms/models/sms_sms.py

```diff
@@ -140,34 +146,26 @@ def resend_failed(self):
 
     @api.model
     def _process_queue(self, ids=None):
-        """ Send immediately queued messages, committing after each message is sent.
-        This is not transactional and should not be called during another transaction!
+        """ CRON job to send queued SMS messages.
 
-       :param list ids: optional list of emails ids to send. If passed no search
-         is performed, and these ids are used instead.
+       :param list ids: optional list of emails ids to send. If passed search
+         only for these records.
         """
         domain = [('state', '=', 'outgoing'), ('to_delete', '!=', True)]
+        if ids is not None:
+            warnings.warn("Since 19.0, ids is deprecated", DeprecationWarning)
+            domain.append(('id', 'in', ids))
 
-        filtered_ids = self.search(domain, limit=10000).ids  # TDE note: arbitrary limit we might have to update
-        if ids:
-            ids = list(set(filtered_ids) & set(ids))
-        else:
-            ids = filtered_ids
-        ids.sort()
+        batch_size = self._get_send_batch_size()
+        records = self.search(domain, limit=batch_size, order='id').try_lock_for_update()
+        if not records:
+            return
 
-        res = None
-        try:
-            # auto-commit except in testing mode
-            auto_commit = not modules.module.current_test
-            res = self.browse(ids).send(unlink_failed=False, unlink_sent=True, auto_commit=auto_commit, raise_exception=False)
-        except Exception:
-            _logger.exception("Failed processing SMS queue")
-        return res
-
-    def _split_batch(self):
-        batch_size = int(self.env['ir.config_parameter'].sudo().get_param('sms.session.batch.size', 500))
-        for sms_batch in tools.split_every(batch_size, self.ids):
-            yield sms_batch
+        records._send(unlink_failed=False, unlink_sent=True, raise_exception=False)
+        self.env['ir.cron']._commit_progress(len(records), remaining=0 if len(records) == batch_size else self.search_count(domain))
```

Comment:
Isn't it the other way around? There's none remaining only if there were fewer records than the batch size?